PYME/LMVis/layers/octree.py

```diff
@@ -68,106 +69,77 @@ def update_from_datasource(self, ds):
-
-            #
-            #     z
-            #     ^
-            #     |
-            #    v1 ----------v3
-            #    /|           /|
-            #   / |          / |
-            #  v5----------v7  |
-            #  |  |    c    |  |
-            #  | v0---------|-v2
-            #  | /          | /
-            #  v4-----------v6---> y
-            #  /
-            # x
-            #
-            # Now note that the counterclockwise triangles (when viewed straight-on) formed along the faces of the cube are:
-            #
-            # v0 v2 v6
-            # v0 v4 v5
-            # v1 v3 v2
-            # v2 v0 v1
-            # v3 v1 v5
-            # v3 v7 v6
-            # v4 v6 v7
-            # v5 v1 v0
-            # v5 v7 v3
-            # v6 v2 v3
-            # v6 v4 v0
-            # v7 v5 v4
-
-
-            # Concatenate vertices, interleave, restore to 3x(3N) points (3xN triangles),
-            # and assign the points to x, y, z vectors
-            triangle_v0 = np.vstack((v0, v0, v1, v2, v3, v3, v4, v5, v5, v6, v6, v7))
-            triangle_v1 = np.vstack((v2, v4, v3, v0, v1, v7, v6, v1, v7, v2, v4, v5))
-            triangle_v2 = np.vstack((v6, v5, v2, v1, v5, v6, v7, v0, v3, v3, v0, v4))
-
-            x, y, z = np.hstack((triangle_v0, triangle_v1, triangle_v2)).reshape(-1, 3).T
-
-            # Now we create the normal as the cross product (triangle_v2 - triangle_v1) x (triangle_v0 - triangle_v1)
-            triangle_normals = np.cross((triangle_v2 - triangle_v1), (triangle_v0 - triangle_v1))
-            # We copy the normals 3 times per triangle to get 3x(3N) normals to match the vertices shape
-            xn, yn, zn = np.repeat(triangle_normals.T, 3, axis=1)
+            shifts = (box_sizes[:,None]*OCT_SHIFT[None,:])*0.5
+            v = (c[:,None,:] + shifts)
+
+            # Counterclockwise triangles (when viewed straight-on) formed along
+            # the faces of an octree box
+            t0 = np.vstack(v[:,[0,0,0,0,0,1,1,1,2,2,4,5],:])
+            t1 = np.vstack(v[:,[2,1,5,6,4,2,3,7,6,7,5,7],:])
+            t2 = np.vstack(v[:,[1,5,4,2,6,3,7,5,7,3,6,6],:])
+
+            x, y, z = np.hstack([t0,t1,t2]).reshape(-1, 3).T  # positions
+
+            tn = np.cross((t2-t1),(t0-t1))
+            xn, yn, zn = np.repeat(tn.T, 3, axis=1)  # normals
+
+            # Color is fixed constnat for octree
+            c = np.ones(len(x))
+            clim = [0, 1]
             
             alpha = self.alpha*alpha/alpha.max()
             alpha = (alpha[None,:]*np.ones(12)[:,None])
             alpha = np.repeat(alpha.ravel(), 3)
             print('Octree scaled alpha range: %g, %g' % (alpha.min(), alpha.max()))
 
-            # Pass the restructured data to update_data
-            self.update_data(x, y, z, cmap=getattr(cm, self.cmap), clim=self.clim, alpha=alpha, xn=xn, yn=yn, zn=zn)
+            cmap = getattr(cm, self.cmap)
```

Comment:
I guess this is because update_data is now missing from TriangleRenderLayer? If so, then that is likely a problem.

Reply:
It looks like `update_data` was deleted in b74fea88e6ac5236c9d526768db4d59c73b87914. Hasn't been an issue.

Reply:
My rationale for calling it a problem was having essentially the same code duplicated in different places, which is never a good thing - I suspect I refactored it without realising that it was also used here.